src/main/scala/org/broadinstitute/dsde/workbench/sam/openam/PostgresAccessPolicyDAO.scala

```diff
@@ -798,6 +842,59 @@ class PostgresAccessPolicyDAO(protected val dbRef: DbReference,
     })
   }
 
+  override def listUserResourceActions(resourceId: FullyQualifiedResourceId, user: WorkbenchUserId, samRequestContext: SamRequestContext): IO[Set[ResourceAction]] = {
+    runInTransaction("listAccessPoliciesForUser", samRequestContext)({ implicit session =>
+      val userPoliciesCommonTableExpression = userPoliciesOnResourceCommonTableExpressions(resourceId, user)
+
+      val userResourcePolicyTable = userPoliciesCommonTableExpression.userResourcePolicyTable
+      val userResourcePolicy = userPoliciesCommonTableExpression.userResourcePolicy
+      val cteQueryFragment = userPoliciesCommonTableExpression.queryFragment
+
+      val policyRole = PolicyRoleTable.syntax("policyRole")
+      val resourceRole = ResourceRoleTable.syntax("resourceRole")
+      val policyActionJoin = PolicyActionTable.syntax("policyActionJoin")
+      val policyAction = ResourceActionTable.syntax("policyAction")
+      val roleActionJoin = RoleActionTable.syntax("roleActionJoin")
+      val roleAction = ResourceActionTable.syntax("roleAction")
+
+      val listUserResourceActionsQuery = samsql"""$cteQueryFragment
+        select ${roleAction.action} as action
+          from ${userResourcePolicyTable as userResourcePolicy}
+          join ${PolicyRoleTable as policyRole} on ${userResourcePolicy.policyId} = ${policyRole.resourcePolicyId} and ${userResourcePolicy.inherited} = ${policyRole.descendantsOnly}
+          join ${ResourceRoleTable as resourceRole} on ${policyRole.resourceRoleId} = ${resourceRole.id} and ${userResourcePolicy.baseResourceTypeId} = ${resourceRole.resourceTypeId}
+          join ${RoleActionTable as roleActionJoin} on ${resourceRole.id} = ${roleActionJoin.resourceRoleId}
+          join ${ResourceActionTable as roleAction} on ${roleActionJoin.resourceActionId} = ${roleAction.id}
+        union
+        select ${policyAction.action} as action
+          from ${userResourcePolicyTable as userResourcePolicy}
+          join ${PolicyActionTable as policyActionJoin} on ${userResourcePolicy.policyId} = ${policyActionJoin.resourcePolicyId} and ${userResourcePolicy.inherited} = ${policyActionJoin.descendantsOnly}
+          join ${ResourceActionTable as policyAction} on ${policyActionJoin.resourceActionId} = ${policyAction.id} and ${userResourcePolicy.baseResourceTypeId} = ${policyAction.resourceTypeId}"""
+
+      listUserResourceActionsQuery.map(rs => ResourceAction(rs.string("action"))).list().apply().toSet
```

Comment:
I had tried to extract it but it is too hard since the string is used both as a string here but also embedded in the sql interpolated string and trying to include the string within the interpolation wants to make it a bind parameter which it isn't. So, yes, magic string, but I could not find a good way around it.